### DISCUSSION TITLE
how to use rpc providers with walletClient

### DISCUSSION THREAD

**Discussion by brossetti1** [wevm/viem#920] [Question]
we are currently doing an ethers to viem migration of our app and I am running into a situation i do not completely understand so I figured id post this here to get a clearer answer.

when we `createWalletClient` with a http transport using `https://our-rpc-provider-url` (ive tried infura, rivet, and default rpc provider for goerli), our `writeContract` calls fail with one of the these errors depending on weather we pass an rpc provider to http transport or not -- ie `http()`:

```
TransactionExecutionError: An error occurred while executing: the method is currently not implemented: eth_sendTransaction
```

it seems like from the docs (https://viem.sh/docs/clients/wallet.html#optional-extend-with-public-actions) that this should work. Also, when we use EIP-1193 `custom(window.ethereum)` as the transport, this `writeContract` call works fine. 

1) does the optional-extend-with-public-actions example only work with the `http` transport because the account is being loaded from private key? 

2) or is it that when we are queueing transactions from a wallet account we should specifically use window.ethereum as the transport?

can someone help me understand what knowledge gaps I am missing, I am assuming it is something simple regarding http transport vs custom(window.ethereum) but I cannot figure it out or find a clear answer anywhere in the issues. thanks for any help you can provide

here is the code in question (some psuedocode)

```
import {
  createWalletClient,
  http,
  PublicClient
} from 'viem';

export async function prepareTX({
  publicClient: PublicClient,
  address: `0x${string}`,
  processedArgs: any
}) => {
    const [account] = await window.ethereum.request({
      method: 'eth_requestAccounts',
    });

    const walletClient = createWalletClient({
      chain: goerli,
      account,
      transport: http('https://our-rpc-provider-url')
    })

    const { request } = await publicClient.simulateContract({
      account: account,
      address: address,
      abi,
      args: processedArgs,
      functionName: method,
    });

    // fails with: method is currently not implemented: eth_sendTransaction 
    // works when we use custom(window.ethereum) as the transport
    const txHash = await walletClient.writeContract(request);
    ...
}
```

ps if we try to get the account with `walletClient.getAddresses` the return value is `undefined` so I have kept the `window.ethereum.request` here for example purposes.

**Comment by summerisgood1** [comment]
Did you use http transport using ``https://our-rpc-provider-url `` for write contract in ether.js instead of ``window.ethereum``?
Since I always use ``https://our-rpc-provider-url `` for read contract only.

**Reply by brossetti1** [reply]
yeah so i think there is just a fundamental misunderstanding on my part in regards to rpc provider vs `window.ethereum` 

can anyone confirm some statements I am trying to confirm:

1) rpc providers host full nodes and only allow you to read the history of the chain - ie to your point, you cannot use rpc providers for writing to the chain, correct?

2) window.ethereum is for interacting with wallets, **including Gnosis Safe's** -- my understanding before this endeavor was that  `window.ethereum` was native to metamask specfically, but after reading up on **eip 1193**, it seems like **all** wallet providers use `window.ethereum` to inject their own wallet interactions into the browser. If so, does this also apply to interacting with gnosis safe through wallet connect or gnosis safe apps as well?

**Reply by OoXooOx** [reply]
It's nonsense. They are provide usage bundle wagmi+web3modal+walletconnect+viem, but in docs they are give example for window.ethereum and no example how fetch transport from web3modal and use it in createWalletClient from viem. I wasted so many time for that...

**Reply by brossetti1** [reply]
@OoXooOx do you have any reference to other docs or example on how to do what you're saying? ie get the transport from web3modal?

are you talking about using the `w3mConnector` in the wagmi config?

```
const wagmiConfig = createConfig({
  autoConnect: true,
  connectors: w3mConnectors({ projectId, chains }),
  publicClient
})
```

**Comment by OoXooOx** [comment]
You can read my theme, mb you can fetch something valuable from it https://github.com/wagmi-dev/viem/discussions/960.

**Comment by cryptostar7** [comment]
I have some problem in using window.ethereum.

When I use following code:
const ownerActionClient = createWalletClient({
      chain: sepolia,
      transport: custom(window.ethereum)
}),

This code occurs error which is that "Argument of type 'WindowProvider | undefined' is not assignable to parameter of type 'EthereumProvider'.
  Type 'undefined' is not assignable to type 'EthereumProvider'."
  
  What is the reasona and how can I fix it?

**Reply by jxom** [reply]
The reason is because `window` is undefined in server environments... A workaround would be to guard client creation in `if (typeof window !== 'undefined')`, or create a "noop" EIP-1193 provider:

```tsx
const noopProvider = { request: () => null } as unknown as EIP1193Provider
const provider = typeof window !== 'undefined' ? window.ethereum! : noopProvider
const client = createWalletClient({
  chain: mainnet,
  transport: custom(provider),
})
```

**Reply by cryptostar7** [reply]
Thank you very much. I used your advice to fix the error. Thank you again
